### PR TITLE
Push footer to bottom and fix content vertically centered and fix scroll to top

### PR DESCRIPTION
**Pull request by RickR2H**
Pull Request for Issue #35003 and replacement of PR #35004

### Summary of Changes
Added holy grail layout to the Cassiopeia template. That way the footer can be pushed to the bottom if there is less content then the screen height. Changed the way the scroll to top is implemented. In the process also removed some obsolete CSS code.

### Testing Instructions
See issue #35003 and description of PR #34849
Please test with and without modules and content and in different display sizes. Also test with short content and some module in footer and header positions.

Don't forget to test the error page as well. This can be done bij typing a URL which doesn't exist on the frontend of the site. For instance https://[test site url]/index.php/thislinkdoesnotexist. This will generate a 404 error, page not found. Make sure the footer is also be on the bottom of the site.

### CODE REVIEW

**Comment by RickR2H** [comment]
@drmenzelit @hans2103 @dgrammatiko @richard67 please review / test.

File: templates/cassiopeia/index.php [hunk 1]
```diff
@@ -166,57 +166,59 @@
 		<?php endif; ?>
 	</header>
 
-	<?php if ($this->countModules('banner', true)) : ?>
-		<div class="container-banner full-width">
-			<jdoc:include type="modules" name="banner" style="none" />
-		</div>
-	<?php endif; ?>
+    <div class="site-grid">
```
Comment (richard67) [inline templates/cassiopeia/index.php]:
```suggestion
	<div class="site-grid">
```
Fix code style error which causes drone to fail.

Please use tabs and not spaces to indent PHP code.

**Comment by hans2103** [comment]
@RickR2H can you move the fix for scroll-to-top to another PR or adjust the description and title of this PR?
You're changing more than you describe.

File: templates/cassiopeia/index.php (hunk 1 above)
Comment (RickR2H) [reply templates/cassiopeia/index.php]: @richard67 sorry. Pushed fix.

**Comment by RickR2H** [comment]
@hans2103 All changes are connected, so I changed the title and description.

**Comment by hans2103** [comment]
please adjust the following: 
```
templates/cassiopeia/scss/blocks/_back-to-top.scss
7:3  ✖  Expected "position" to come before "opacity"     order/properties-order
9:3  ✖  Expected "right" to come before "bottom"         order/properties-order
17:3  ✖  Expected "z-index" to come before "transition"   order/properties-order
```


that's why the build fails

**Comment by richard67** [comment]
@RickR2H Regarding your review request: I think I'm not the right guy to review since my knowledge on CSS and JS is too little.

**Comment by RickR2H** [comment]
> please adjust the following:
> 
> ```
> templates/cassiopeia/scss/blocks/_back-to-top.scss
> 7:3  ✖  Expected "position" to come before "opacity"     order/properties-order
> 9:3  ✖  Expected "right" to come before "bottom"         order/properties-order
> 17:3  ✖  Expected "z-index" to come before "transition"   order/properties-order
> ```
> 
> that's why the build fails

Thanks Hans. Pushed new commit

**Comment by richard67** [comment]
Just wanted to post but Hans was faster.

**Comment by hans2103** [comment]
@RickR2H onwards to solve the next check. 
```
	templates/cassiopeia/scss/blocks/_back-to-top.scss
8	  8:3  ✖  Expected "right" to come before "opacity"           order/properties-order
9	 16:3  ✖  Expected "z-index" to come before "border-radius"   order/properties-order
```

note... if you only move `right`, the next build will fail because `bottom` needs to be before `opacity` too.
You might consider using scss code styling check in your IDE.

**Comment by richard67** [comment]
@RickR2H You can use following commands to check code style on the command line if you have npm:

`npm run lint:css`

`npm run lint:js`

**Comment by RickR2H** [comment]
> npm run lint:css

Thanks Richard! I'll do that next time!

**Comment by hans2103** [comment]
@RickR2H I see a lot of spacing added... is that correct?
https://github.com/joomla/joomla-cms/pull/35012/files
Schermafbeelding 2021-08-02 om 11 42 42

**Comment by RickR2H** [comment]
Mmmm that's look like something is, wrong with the indentation. I'll try to fix today.

**Comment by RickR2H** [comment]
@hans2103 Did a review of the whole index.php to make indentation more structured. To review the indentations, please take a look at the whole file instead of looking at only the changes in Git. Hope these changes are more appealing to the eye 😉

**Comment by brianteeman** [comment]
Something still not correct

https://github.com/joomla/joomla-cms/blob/78f11a01f2fb09bb0b534e54476f2e452556bd1b/templates/cassiopeia/index.php

**Comment by RickR2H** [comment]
@brianteeman @hans2103 Is it a better option to remove the empty lines?

please review https://github.com/joomla/joomla-cms/blob/d374cbdb98bdca5e6869d6120c8140d6e34f9a62/templates/cassiopeia/index.php

**Comment by dgrammatiko** [comment]
@RickR2H please remove the `.gz` file

File: templates/cassiopeia/scss/blocks/_back-to-top.scss [hunk 2]
```diff
@@ -1,36 +1,24 @@
...
-.back-to-top-wrapper {
-  position: absolute;
-  top: calc(#{$scrollLength} + #{$cassiopeia-grid-gutter * 10});
-  right: 0;
-  bottom: 0;
-  z-index: $zindex-fixed;
-  width: 3rem;
-  pointer-events: none;
-
-  [dir=rtl] & {
```
Comment (hans2103) [inline templates/cassiopeia/scss/blocks/_back-to-top.scss:16]: @RickR2H I think this part should be added to className .back-to-top-link too

**Review by hans2103 — CHANGES_REQUESTED** [review]
@RickR2H You're removing dir=ltr from the back-to-top-wrapper but not implementing it back to back-to-top-link. I think you should add it to back-to-top-link too

File: templates/cassiopeia/scss/blocks/_back-to-top.scss (hunk 2 above)
Comment (RickR2H) [reply templates/cassiopeia/scss/blocks/_back-to-top.scss:16]: Good catch! Forgot that part...

Comment (RickR2H) [reply templates/cassiopeia/scss/blocks/_back-to-top.scss:16]: Commit added

**Review by hans2103 — APPROVED** [review]

**Comment by hans2103** [comment]
I have tested this item :white_check_mark: successfully on daf07c2c83e692c3417b6a1c12aa0fc9ffd00de0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35012">issues.joomla.org/tracker/joomla-cms/35012</a>.</sub>

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on daf07c2c83e692c3417b6a1c12aa0fc9ffd00de0<br /><br />Tested with and without content, with and without back-to-top, with and without modules on the sidebar, in rtl and in different display sizes.

The case with very short content and some modules in bottom positions is a very edge case I guess, so I can live with bottom modules not being at the very bottom of the page ;-)

Thank you @RickR2H  for your work.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35012">issues.joomla.org/tracker/joomla-cms/35012</a>.</sub>

**Comment by RickR2H** [comment]
@drmenzelit @hans2103 thanks for testing<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35012">issues.joomla.org/tracker/joomla-cms/35012</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35012">issues.joomla.org/tracker/joomla-cms/35012</a>.</sub>

**Comment by RickR2H** [comment]
@wilsonge dm on Glip

**Comment by PhilETaylor** [comment]
@wilsonge This works and fixes #35119. LGTM :)

**Comment by PhilETaylor** [comment]
Actually... while this fixes the immediate issue of #25119 enough to launch Joomla 4, the 403 error page is screwed up and the footer appears above the content

Screenshot 2021-08-16 at 22 11 30

**Comment by softforge** [comment]
I have tested this item :white_check_mark: successfully on daf07c2c83e692c3417b6a1c12aa0fc9ffd00de0<br /><br />Further tests on this and still its working fine<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35012">issues.joomla.org/tracker/joomla-cms/35012</a>.</sub>

**Comment by PhilETaylor** [comment]
> Further tests on this and still its working fine 

Except its not though.  Test the error page. The 404 page. The 403 error page.

**Comment by RickR2H** [comment]
@PhilETaylor I can check if this can be fixed really quickly...

File: templates/cassiopeia/js/template.js [hunk 3]
```diff
@@ -37,7 +37,23 @@ Joomla = window.Joomla || {};
      * Back to top
      */
     var backToTop = document.getElementById('back-top');
+
     if (backToTop) {
+
+      function checkScrollPos() {
+        if (document.body.scrollTop > 20 || document.documentElement.scrollTop > 20) {
```
Comment (PhilETaylor) [inline templates/cassiopeia/js/template.js:44]: Just being picky but no body needs a "back to top" link box after just 20 pixels of scrolling ! haha

Comment (brianteeman) [reply templates/cassiopeia/js/template.js:44]: and thats why its optional ;)

Comment (richard67) [reply templates/cassiopeia/js/template.js:44]: I'm pretty sure we will get some issues "just because" ;-)

Comment (PhilETaylor) [reply templates/cassiopeia/js/template.js:44]: At least make it when the height of the viewport +1pixel... else its meaningless... no one needs to click a link to view the first 20 pixels of  screen then just scrolled down 20px on...

**Comment by PhilETaylor** [comment]
also if you have a named link fragment like 

```html
<a href="#top"......
```

but there is no named anchor for `top`, is that even valid? Probably valid HTML but still not perfect right? 

Yes i know JS is doing the "actual" `window.scrollTo` but there should be a named anchor for completeness right?

**Comment by PhilETaylor** [comment]
Also - is it meant to smooth scroll still? because it doesnt, it jumps instantly when clicked.

**Comment by PhilETaylor** [comment]
> Also - is it meant to smooth scroll still? because it doesnt, it jumps instantly when clicked.

Ignore that - Safari on mac doesnt like smooth scroll apparently. Works in google chrome... yet another difference with Joomla and Safari :(

**Comment by RickR2H** [comment]
@PhilETaylor the error page is fixed...

**Comment by RickR2H** [comment]
Problem now is that is need two tests to pass

**Comment by richard67** [comment]
@RickR2H Could you update your testing instructions by adding a short hint to check also the error page? Thanks in advance.

**Comment by PhilETaylor** [comment]
> @PhilETaylor the error page is fixed...

Yes I can confirm its working for me too now - thanks for quickly fixing this. Ive not comprehensively tested the rest of the PR, apart from it fixes my reported issue in #35119 and now fixes the error page too.

@richard67 @wilsonge Im happy if you are, the other tests are valid still I think because the changes only changed the error.php files.

**Comment by richard67** [comment]
@PhilETaylor If you mark your test result in the issue tracker here https://issues.joomla.org/tracker/joomla-cms/35012 and @softforge or one of the other previous testers is available to test again, that would be fine.

**Comment by PhilETaylor** [comment]
I have tested this item :white_check_mark: successfully on ec7881f66ec52e806d976500bfc36fff9ee20ade<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35012">issues.joomla.org/tracker/joomla-cms/35012</a>.</sub>

**Comment by RickR2H** [comment]
Updated the test instructions as mentioned

**Comment by RickR2H** [comment]
@richard67 we still need one more test to pass 😉

**Comment by softforge** [comment]
I have tested this item :white_check_mark: successfully on ec7881f66ec52e806d976500bfc36fff9ee20ade<br /><br />Had server errors on. Switched off and could confirm Phil T issue. Updated to the latest build and confirm the footer switched position to the correct position on 404 and 403.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35012">issues.joomla.org/tracker/joomla-cms/35012</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Comment by richard67** [comment]
Has anyone tested RTL?

**Comment by richard67** [comment]
I am just testing, too, and just wanted to.

**Comment by PhilETaylor** [comment]
> Has anyone tested RTL?

I cant even spell my name backward. So no.

**Comment by RickR2H** [comment]
Will check now...

**Comment by richard67** [comment]
RTL looked ok here, the rest too.

I've noticed that the error page doesn't respect the "Sticky Header" option of the template style settings, but that might also have been the case without this PR. am just checking that.

**Comment by richard67** [comment]
Yes, was also case before this PR. So I have not found any issues, all ok.

Now let's hope nobody of us missed anything ;-)

**Comment by RickR2H** [comment]
@richard67 RTL works! Also tested the error page. Thanks to all!

**Pull request merged**